core-rust/state-manager/src/store/rocks_db.rs

```diff
@@ -1996,6 +2059,104 @@ impl<R: WriteableRocks> StateManagerDatabase<R> {
+            let type_info_creation = substate_changes.get(
+                node_id,
+                &TYPE_INFO_FIELD_PARTITION,
+                &TypeInfoField::TypeInfo.into(),
+            );
+            let Some(type_info_creation) = type_info_creation else {
+                continue;
+            };
+            let SubstateChangeAction::Create { new } = type_info_creation else {
+                panic!(
+                    "type info substate should be immutable: {:?}",
+                    type_info_creation
+                );
+            };
+            let type_info = scrypto_decode::<TypeInfoSubstate>(new).expect("decode type info");
+
+            let entity_type = node_id.entity_type().expect("type of upserted ReNode");
+            let creation_id = CreationId::new(state_version, index_within_txn);
+
+            match type_info {
+                TypeInfoSubstate::Object(object_info) => {
+                    let blueprint_id = object_info.blueprint_info.blueprint_id;
+                    let BlueprintId {
+                        package_address,
+                        blueprint_name,
+                    } = blueprint_id.clone();
+                    db_context.cf(TypeAndCreationIndexedEntitiesCf).put(
+                        &(entity_type, creation_id.clone()),
+                        &EntityBlueprintId::of_object(*node_id, blueprint_id),
+                    );
+                    db_context.cf(BlueprintAndCreationIndexedObjectsCf).put(
+                        &(package_address, hash(&blueprint_name), creation_id),
+                        &ObjectBlueprintNameV1 {
+                            node_id: *node_id,
+                            blueprint_name,
+                        },
+                    );
+                }
+                TypeInfoSubstate::KeyValueStore(_kv_store_info) => {
+                    db_context.cf(TypeAndCreationIndexedEntitiesCf).put(
+                        &(entity_type, creation_id),
+                        &EntityBlueprintId::of_kv_store(*node_id),
+                    );
+                }
+                TypeInfoSubstate::GlobalAddressReservation(_)
+                | TypeInfoSubstate::GlobalAddressPhantom(_) => {
+                    panic!("should not be persisted: {:?}", type_info)
+                }
+            }
+        }
+    }
+
+    fn catchup_entity_listing_indices(&self) {
+        const TXN_FLUSH_INTERVAL: u64 = 10_000;
+
+        info!("ReNode listing indices are enabled.");
+        let db_context = self.open_rw_context();
+        let catchup_from_version = db_context
+            .cf(ExtensionsDataCf)
+            .get(&ExtensionsDataKey::EntityListingIndicesLastProcessedStateVersion)
+            .map(StateVersion::from_be_bytes)
+            .unwrap_or(StateVersion::pre_genesis())
+            .next()
+            .expect("next version");
+
+        let mut receipts_iter = db_context
```

Comment:
Hmm I thought `catchup_from_version` is actually `previous_indexed_up_to_version`?
So shouldn't we start iterating from `previous_indexed_up_to_version + 1` inclusive?

Reply:
Although I imagine setting these is idempotent, so we won't have duplicated entries? So it might be safe?

Reply:
a tiny `.next()` (in line 2132 above) is the `+ 1` you are looking for :D

FOR ONCE I avoided the "off by one" 🎉 (I think 🤔)

Reply:
Ahh I thought that was a `.next()` to return the first value from an iterator! 
In retrospect, perhaps `next` is a little too overloaded, but it's minor anyway?

In terms of this code, tiny suggestion - maybe separating out `let catchup_from_version  = previously_updated_to_version.next();` might make it clearer / more explicit for noobs like me?

Reply:
this I can do: https://github.com/radixdlt/babylon-node/pull/926/commits/e8b08ccc5ff4acb8877f003d295c97070598743f 👍